pyquil/api/quantum_computer.py

```diff
@@ -0,0 +1,183 @@
+        """
+        Queue a quil program for running, but return immediately with a job id.
+
+        Use :py:func:`QuantumComputer.wait_for_job` to get the actual job results, probably
+        after queueing up a whole batch of jobs.
+
+        See :py:func:`run` for this function's parameter descriptions.
+
+        :returns: a job id
+        """
+        if symmetrize is None:
+            symmetrize = self.symmetrize
+
+        if not classical_addresses:
+            classical_addresses = get_classical_addresses_from_program(program)
+
+        if symmetrize:
+            raise NotImplementedError("Async symmetrized measurement isn't supported")
+
+        return self.qam.run_async(program, classical_addresses, trials)
+
+    def run_symmetrized(self, program, classical_addresses, trials):
+        """
+        Run a quil program in such a way that the readout error is made collectively symmetric
+
+        This means the probability of a bitstring ``b`` being mistaken for a bitstring ``c`` is
+        the same as the probability of ``not(b)`` being mistaken for ``not(c)``
+
+        A more general symmetrization would guarantee that the probability of ``b`` being
+        mistaken for ``c`` depends only on which bit of ``c`` are different from ``b``. This
+        would require choosing random subsets of bits to flip.
+
+        In a noisy device, the probability of accurately reading the 0 state might be higher
+        than that of the 1 state. This makes correcting for readout more difficult. This
+        function runs the program normally ``(trials//2)`` times. The other half of the time,
+        it will insert an ``X`` gate prior to any ``MEASURE`` instruction and then flip the
+        measured classical bit back.
+
+        See :py:func:`run` for this function's parameter descriptions.
+        """
+        flipped_program = _get_flipped_protoquil_program(program)
+        if trials % 2 != 0:
+            raise ValueError("Using symmetrized measurement functionality requires that you "
+                             "take an even number of trials.")
+        half_trials = trials // 2
+
+        samples = self.run(program, classical_addresses, half_trials, symmetrize=False)
+        flipped_samples = self.run(flipped_program, classical_addresses, half_trials,
+                                   symmetrize=False)
+        double_flipped_samples = np.logical_not(flipped_samples).astype(int)
+        results = np.concatenate((samples, double_flipped_samples), axis=0)
+        np.random.shuffle(results)
+        return results
+
+    def run_and_measure(self, program: Program, qubits: List[int], trials: int, symmetrize=None):
+        """
+        Automatically apply MEASURE instructions to each qubit.
+
+        I think we should discourage use of this function. ``run`` is much more general.
+
+        Unlike the old ``run_and_measure`` function, this one works with noise models.
```

Comment:
I'm fine with this as long as we include measure_all(). Then people could just do:
```
prog.measure_all()
qpu.run(prog)
# instead of
qpu.run_and_measure(prog)
```

Reply:
changed my mind. `run_and_measure` is fine if it's just a convenience wrapping for adding measure and calling run (as it is now). The pain was in how the behavior differs between QVMConnection and QPUConnection (esp wrt noise models). Updated the docstrings